tutorials/math/matrix-exponentiation.md

```diff
@@ -9,8 +9,522 @@ keywords:
+
+Now, we have a linear recurrence. For Let us try to find the matrix $T$ such that 
+$$$
+T \cdot
+\begin{bmatrix}
+  f(len - 1, a) \\
+  f(len - 1, e) \\
+  f(len - 1, i) \\
+  f(len - 1, o) \\
+  f(len - 1, u) 
+\end{bmatrix}
+= 
+\begin{bmatrix}
+  f(len, a) \\
+  f(len, e) \\
+  f(len, i) \\
+  f(len, o) \\
+  f(len, u) 
+\end{bmatrix}
+= 
+\begin{bmatrix}
+  \begin{array}{l}
+    f(len - 1, e) + f(len - 1, i) + f(len - 1, u)\\
+    f(len - 1, a) + f(len - 1, i)\\
+    f(len - 1, e) + f(len - 1, o)\\
+    f(len - 1, i) \\
+    f(len - 1, i) + f(len - 1, o)
+  \end{array}
+\end{bmatrix}
+$$$
+
+<details>
+
+<summary> Finding T </summary>
+
+$$$
+T = 
+\begin{bmatrix}
+0 & 1 & 1 & 0 & 1 & \\ 
+1 & 0 & 1 & 0 & 0 & \\ 
+0 & 1 & 0 & 1 & 0 & \\ 
+0 & 0 & 1 & 0 & 0 & \\ 
+0 & 0 & 1 & 1 & 0 & 
+\end{bmatrix}
+$$$
+
+If you are unable to see how we came up with matrix $T$, then for simplicity let us write $f(len - 1, ch)$ as only $ch$. Now $T \cdot A = B$ looks as shown below:
+
+$$$
+T \cdot \begin{bmatrix}
+a \\
+e \\
+i \\
+o \\ 
+u
+\end{bmatrix} 
+= 
+\begin{bmatrix}
+    e + i + u \\
+    a + i \\
+    e + o \\
+    i \\
+    i + o
+\end{bmatrix}
+$$$
+
+</details>
+
+Our `multiply` function has one change that was mentioned at the end of last problem (taking modulo at each operation) and the `power` function remains unchanged.
+
+<Tabs>
+<TabItem value="cpp" label="C++">
+<SolutionAuthor name="@Ishwarendra"/>
+
+```cpp
+int countVowelPermutation(int n) {
+  std::vector T(5, std::vector<int>(5, 0));
+  T[0][1] = T[0][2] = T[0][4] = 1;
+  T[1][0] = T[1][2] = 1;
+  T[2][1] = T[2][3] = 1;
+  T[3][2] = 1;
+  T[4][2] = T[4][3] = 1;
+
+  // There is 1 way to make string of length-1 whose ending character is fixed.
+  std::vector A(5, std::vector(1, 1)); 
+
+  // Since our base case is string of length-1 so we need T^(n - 1) only to get answer for string of length n = (1 + (n - 1))
+  auto B = multiply(power(T, n - 1), A);
+
+  int ans = 0;
+  for (int i = 0; i < std::size(B); i++) {
+    ans = (ans + B[i][0]) % MOD;
+  }
+
+  return ans;
+}
+```
+
+</TabItem>
+</Tabs>
+
+- **Time Complexity:** $O(log(n)$
+- **Space Complexity:** $O(1)$ as we have created 2D vector of constant dimension only.
+
+#### General Advice:
+- You can use matrix exponentiation only if the recurrence is linear and has constant coefficients.
+- Matrix exponentiation can often be used as an optimisation for dynamic programming solutions. Again check if the recurrence formed is **linear with constant coefficients or not!.**
+
+export const suggestedProblems = [
+  {
+    "problemName": "1137. N-th Tribonacci Number",
+    "difficulty": "Easy",
+    "leetCodeLink": "https://leetcode.com/problems/n-th-tribonacci-number/description/",
+    "solutionLink": "../../../solutions/1100-1199/1137-n-th-tribonacci-number-easy.md"
```

Comment:
this link doesn't exist.

Reply:
Maybe the `.md` at end is causing the problem 🤔. Although similar link worked in case of binary exponentiation.

Reply:
Found the problem, we don't need to write problem number. Fixed it.